### ISSUE TITLE
Support AppImages

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/itch#683]
http://appimage.org/ — this is what Krita uses for their Linux distributions

**Comment by probonopd** [comment]
Here is a recipe that does it. But I didn't test the app since the first thing it does is force me to sign up somewhere, which is too much of a hurdle for me.

https://github.com/probonopd/AppImages/blob/master/recipes/meta/Itch.yml

Remaining issues to be solved (FIXME):
- Apparently keeps running forever, even after closing the last window (seems unrelated to AppImage)
- Checks https://nuts.itch.zone/update/linux_x64/17.6.6 - if this is some sort of auto-updater, then keep in mind that AppImages are intentionally read-only so that they can't modify themselves. But have a look at AppImageUpdate for binary delta updates. This could be especially useful for Electron-based apps since Electron probably doesn't change much from build to build.

**Comment by fasterthanlime** [comment]
Since itch downloads and runs applications, this issue was originally opened to support launching applications in the AppImage format. Actually, it already does, since AppImages are already ELF executables.

However, I still think it's a relevant issue, because I'm looking into firejail for #670, and it has a dedicated --appimage flag now, so itch will have to detect when to use it.

Also, the more I'm reading about AppImage, the more I think it's the best solution for distributing itch on Linux desktops in the future. I've taken a look at AppImageUpdate and while it's interesting, I would probably want to implement a transport based on https://github.com/itchio/butler instead!

**Comment by probonopd** [comment]
Thanks for the clarification. Let me know if I can assist you in making that happen.

I've updated the recipe (above) and uploaded an experimental AppImage of the itch client for your testing (some fine-tuning might still be required):

https://bintray.com/probono/AppImages/itch#files

Would be trilled to see AppImage used for
1. Distributing the itch client
2. Distributing the games downloaded/managed by the itch client, with delta updates

**Comment by probonopd** [comment]
@fasterthanlime regarding AppImageUpdate, try, for example, to update [RetroArch-1.3.6+r444.glibc2.17-x86_64.AppImage](https://bintray.com/probono/AppImages/download_file?file_path=RetroArch-1.3.6%2Br444.glibc2.17-x86_64.AppImage) to the latest version. The AppImage includes all the cores and is 229 MB in size, yet the binary delta update to the latest version happens within seconds and downloads only a tiny fraction of the code.

Just launch [AppImageUpdate](https://bintray.com/probono/AppImages/AppImageUpdate/_latestVersion#files) and select the AppImage package you want to update.

What advantages would you see in using something different instead?

**Comment by probonopd** [comment]
If I understand it correctly, itch currently downloads archives like `tiled-2016.05.25-linux64.tar.gz` which are self-contained "almost" [AppDirs](http://rox.sourceforge.net/desktop/AppDirs.html) and unpacks these.

It would be quite straightforward to do the exact same thing but with AppImages instead of .tar.gz files, with the added advantage that you could decide whether you just want to run them (which mounts the image and executes the app inside) or, more traditionally, just unpack the AppImage (exactly as you do with the .tar.gz). An AppImage is an ISO9660 (in the future, squashfs) file, so the unpacking is quite straightforward (e.g., using libarchive, bsdtar, and/or unsquashfs).

**Comment by fasterthanlime** [comment]
> Would be trilled to see AppImage used for
> 1. Distributing the itch client
> 2. Distributing the games downloaded/managed by the itch client, with delta updates

You sound very motivated and I hate to be a downer, but, since you're really asking I'll try to give a complete answer! I'll start with 2., then address 1.

Distributing all games via AppImage is not an option for us for the following reasons:
- At the time of this writing, we have 43K games on itch.io — if we assume an average of 4 games per developers (some have 12, some just the one), that's about 10K people we need to teach about AppImage. We just can't enforce that for everyone.
- For those we _can_ influence (newer games, developers we work more closely with, developers shipping large (>500MB) games), we've developed a system of our own, wharf. The [spec](https://itch.io/docs/wharf/) is a bit bare, so I'd recommend looking at the [reference implementation](https://github.com/itchio/wharf) instead (if you're not allergic to golang,  as many seem to be nowadays!).

Here's a bunch of advantages to wharf in no particular order:
- Diffs are computed once, compressed with brotli-q1, and served as-is — zsync also uses the rsync algorithm (probably with a smaller block size) but we'd need to use HTTP transport compression (gzip, which would incur CPU usage, and still wouldn't come close to the ratios we see with brotli).
- The whole implementation is open, very flexible, and has gotten very reliable. We've been using it in production for a few months now. It doesn't assume any particular platform and as such, works fine on Linux, macOS, and Windows.
- Having all the elements of a diffing/patching/verification/healing system available as a library (which we know in and out) lets us ship great tools: [butler](https://github.com/itchio/butler) is especially good at uploading: taking care of signing, diffing, compression, uploading (even over unreliable connections). All operations have progress meters, detailed logging, etc. Adding proxy support is trivial. Our patch processing backend (which, among other things, generates full .zip archives from old-version+patch for web downloaders) is parallelized and highly-efficient. We're just a few weeks away from deploying a block-based system to make processing and healing even faster, especially for large games (~10GB). It's actually fascinating, I wouldn't mind sharing the details with you if you're curious!

For large games, being able to load directly from disk (without an additional compression layer like AppImage ships with) is essential to performance — load times are a larger issue than disk space usage in the "power game user" scenario.

Lastly, this is Linux. It's a wonder whenever two people agree on any single thing. I'm trying to support as many scenarios as I can without stepping on any toes. All developers/porters have their preferences on how they like to do things (RPATH, LD_LIBRARY_PATH, some sort of container format, etc.) and my job is all about making sure all of these work, without extra work for them.

So, for 2.: like I said, I'd be happy to have great support for AppImage games in the app (including sandbox support, since firejail supports AppImage already!), but it's not something we're going to (or even could) enforce. Our diff/patching system is very well suited to our needs, cross-platform, fast, etc. so there's really no need to fill here.

(Note: you could take each single argument above and take it apart — but trust me when I say: I fully understand how zsync works, and if it was enough for our use-case, I wouldn't have bothered making my own system!)

---

However, 1. is a different story. I've been unhappy about .rpm and .deb for as long as I've had to ship them. Not so long ago I even opened a pretty aggressive issue about it (#827). I'd be really glad to push another "preferred alternative".

The Windows & macOS builds already have packages with the following properties:
- Releases are published on GitHub
- An [update server](http://nuts.itch.zone/notes) (a [nuts](https://github.com/GitbookIO/nuts) instance) reacts to GitHub webhooks to get up-to-date version information and serve file downloads
- The app is installed to a user-controlled location (no need for administrator rights) — on Windows, the installation process is fully transparent, doesn't even show a wizard. (So far so good on the AppImage front)
- Shortcuts are created (by the package), a URL handler is installed (by the app). The app installs a few dependencies of its own (butler, unarchiver, firejail on Linux, elevate+isolate on Windows, activate on macOS — all built and signed by us).
- The app is able to check for updates itself (by querying the update server mentioned above)
- The app is able to download updates while it's running, then when it's done downloading, show a changelog dialog with options "Restart & apply now" or "Snooze")
- When clicking "Restart & apply now", the app quits and starts up again in a matter of seconds (depending on your computer) — no downloading/extracting/patching is done at this step, everything has already been done.

For the macOS version, it's easy (in principle) — running an executable doesn't lock it, so all files in the app bundle can be overwritten while it's launched.

For the Windows version, Squirrel.Windows uses another trick: app versions are stored in numbered folders (`C:\Users\amos\AppData\Local\itch\app-18.10.1`), and the shortcut points to `Update.exe`, which launches the right version.

Whichever solution we eventually go with, for root-less installation and update of itch would have to be at least as good as what I described above (minus the cryptic .NET stacktraces of Squirrel.Windows now and then!).

I think it's achievable using wharf, our diff/patcher — although it would require some level of extra work to integrate with electron's [auto-updater](https://github.com/electron/electron/blob/master/lib/browser/api/auto-updater.js), and so on.

However, since you seem motivated, and if it would help push AppImage on Linux, I'd definitely be willing to look into using AppImage as the recommended way to install itch (and keep alternative packages for all the alternative folks) — again, assuming the final solution fits the requirements above.

**Comment by probonopd** [comment]
Just played around with it a bit, turns out it is easy to turn an itch download into an AppImage. I understand that this might not be feasible for all current downloads on itch, but possibly might be considered for future ones.

```
# Get appimagetool from the last successful build on Travis CI
ID=$(wget -q https://api.travis-ci.org/repos/probonopd/appimagetool/builds -O - | head -n 1 | sed -e 's|}|\n|g' | grep '"result":0' | head -n 1 | sed -e 's|,|\n|g' | grep '"id"' | cut -d ":" -f 2)
URL=$(wget -q "https://s3.amazonaws.com/archive.travis-ci.org/jobs/$((ID+1))/log.txt" -O - | grep "https://transfer.sh/.*/appimagetool" | tail -n 1 | sed -e 's|\r||g')
wget "$URL"
chmod a+x ./appimagetool

# Turn itch.io download into a full AppDir
mkdir Tiled.AppDir
tar xf ./.config/itch/downloads/214445.tar.gz --strip-components=1 -C Tiled.AppDir
cd "Tiled.AppDir"
# The AppImage needs a way to know the main executable, the AppDir spec calls it "AppRun"
ln -s tiled AppRun
# The AppImage needs a .desktop file and icon
wget -c https://github.com/bjorn/tiled/raw/master/tiled.desktop
wget -c https://github.com/bjorn/tiled/blob/master/src/tiled/images/32x32/tiled.png
cd ..

# Finally, make the AppImage
./appimagetool Tiled.AppDir

# Try to run it
./Tiled-x86_64.AppImage
```

**Comment by probonopd** [comment]
What you describe about brotli, butler, and wharf sounds awesome and makes me wonder why we are not using it instead of zsync for all AppImages. Does it work on monolithic files such as zisofs and/or squashfs images?

**Comment by fasterthanlime** [comment]
> Just played around with it a bit, turns out it is easy to turn an itch download into an AppImage. I understand that this might not be feasible for all current downloads on itch, but possibly might be considered for future ones.

I agree it's easy (good job!) and it is interesting to us, because games lacking the proper libraries to run is a big issue on Linux (as you're well-aware). We do have a [guide for that](https://itch.io/docs/itch/integrating/platforms/linux.html) but there's still many games that assume certain libraries are present.

However, it's out of our control - the right folks to petition for this would be developers directly. We can facilitate that (by including it prominently on our docs, reminding developers about it etc.) — once we make sure the AppImage case is well supported in the app.

> What you describe about brotli, butler, and wharf sounds awesome and makes me wonder why we are not using it instead of zsync for all AppImages. Does it work on monolithic files such as zisofs and/or squashfs images?

Well, first off, it's all pretty recent :) it probably didn't exist back when you investigated alternatives.

Second, it does indeed work just as well within files as it does across many different files, but the issue here is compression. As I understand it, zsync special-cases gzip compression, but it does make me wonder what its assumptions are:
- Does it assume it can read the entire old file in memory?
- Does it assume there's enough space on disk for the entire old file and the entire new file?
- Does it require reading the entire old file + writing the entire new file? For small games, it's a non-issue, for large games, it's a big issue (reading 10GB+writing 10GB to disk is no longer negligible, we've learned that the hard way).

I could definitely look into teaching wharf/butler about monolithic gzip-compressed files (or ziofs/squashfs files), but the caveats above would probably still apply (I don't see any way around it currently).

The typical use case for large games is as follows:
- Binaries make up a fairly small portion of the total game contents
- Assets are usually compressed already (.png images, .ogg/.mp3 files, GPU-friendly compressed texture formats, proprietary Unity .assets format, proprietary UE4 level data, etc.)
- Updates often change the binary (which diffs fairly well), and remove assets or add new ones

For the 10GB game I keep mentioning (which has been keeping me up at night for a month+), typical patches are anywhere between 10MB (for a "strictly bug-fix" update) and 250-400MB (for an "additional content" update). However, applying them is especially fast. In the upcoming iteration of our system:
- Downloading, patching, and verification happen in parallel, with low RAM usage (<100MB), since everything is streaming
- Only the re-used parts are read from disk — verification happens before data is written to disk
- A "staging folder" is used so that patch application can be rolled back (by just deleting the staging folder instead of merging it with the installation folder). Disk usage is at most `old version size + changed files size`.
- I'm working on a healing system that would in fact make the staging folder unnecessary, since in case of corruptions, we're not interested in aborting the patching, but instead, redownloading the corrupting parts directly from the newest archive (or block store).

TL;DR: typical patches change small portions of large installs, our solution is focused on doing that quickly, using low RAM, low or no additional disk space, using network and disk bandwidth in parallel, and working even if the old install was corrupted (which happens more than you'd think, especially with large datasets and aging disks).

If you're okay with the caveats above though, we can definitely look into adding zisofs/squashfs support to butler at some point. I still think it would be a valid solution for small games. I can see a few more downsides for your particular use case:
- The butler binary is probably much larger than zsync (~15MB uncompressed as of now — but it does have a lot of other commands, and recent versions of golang are getting better at producing smaller binaries)
- Patches need to be computed somehow. For us, that happens during upload: the uploader downloads the signature (block hashes) of the old version and runs the rsync algorithm on that + the local data (ie. the new version). Then they need to be stored somewhere. zsync works a bit differently: as long as you have the new version fully uploaded somewhere, along with its hashes (the `.zsync` file itself, from what I can see), then you can patch from any version to there. It's a different strategy, equally valid! (although it requires more disk I/O when updating).

**Comment by probonopd** [comment]
I see that you know zsync in and out :-) Indeed not having to pre-computing the deltas between the latest version and every version before was a big decision factor in favor of using zsync for AppImageUpdate. I imagine calculating all the different deltas to be computationally and storage intensive, and I don't have the kind of resources needed for that. Also, AppImage is supposed to be decentral, so that upstream authors can host their own stuff, and for that it must be simple.

I wouldn't be concerned too much about the 15 MB size of butler.

zsync has been in use for Ubuntu Live ISOs for ages; as they are ISO9660 files essentially containing a large squashfs file I was blindly assuming that it can handle these cases pretty well; didn't look much into the technical details (you probably have).

So yes, I'd definitely be interested in squashfs support for butler if we can determine that it will be working well for the use cases you mentioned above.

**Comment by probonopd** [comment]
>  Games lacking the proper libraries to run is a big issue on Linux (as you're well-aware). We do have a guide for that but there's still many games that assume certain libraries are present. However, it's out of our control - the right folks to petition for this would be developers directly.

Do you run any kind of quality control on the uploads? I am asking because the second game I tried to download from itch refuses to run on a stock Ubuntu 16.04... (Love, Money, Rock-n-Roll Demo Linux x64)

**Comment by fasterthanlime** [comment]
> Do you run any kind of quality control on the uploads?

That's definitely on my TODO list! See #375 - it's less simple than it sounds though, for example, ldd allows random code execution, so it needs to be run in a sandbox. And just listing .so files doesn't mean it'll run! I want to do it right when I do it eventually, not interested in half-measures here!

> Also, AppImage is supposed to be decentral, so that upstream authors can host their own stuff, and for that it must be simple.

Exactly, that's why I think zsync is a really good solution for your use case. Except you're saying it doesn't actually support squashfs? It looks like squashfs compress blocks instead of files (or the whole stream), which is much better for diffing (but still needs reading+writing the whole file when upgrading).

**Comment by probonopd** [comment]
Here is an example of an AppImage already being distributed on itch:
https://yousry.itch.io/cool

So, how well does your delta updater perform on this? Can I see the different pre-generated binary deltas?

**Comment by fasterthanlime** [comment]
> Here is an example of an AppImage already being distributed on itch:
> https://yousry.itch.io/cool

There's only one version of this AppImage for this game, we'd have to find a project that actually uses butler to push builds and not use the web uploader! (generating patches for web-uploaded content is also on my list).

**Comment by fasterthanlime** [comment]
It's easy to run offline tests by yourself btw:
- Install butler: https://itch.io/docs/butler/
- Run `butler diff folder1 folder2`
- Look at `patch.pwr`'s size
- You can control brotli compression quality with the `-q` parameter

See `butler help` for other commands (sign, verify, apply, etc.)

Our backend uses the library behind `butler diff` rather than calling the butler binary directly, but it's the same code.

**Comment by bjorn** [comment]
So, apart from supporting the installation of AppImage files and the possible distribution of the itch app as an AppImage, I got another request:

Support uploading AppImage files using `butler`. Currently, it exists with:

```
*errors.errorString zip: not a valid zip file
/home/amos/builds/630f4286/0/itchio/butler/src/github.com/itchio/butler/push.go:372 (0x421ce9)
/usr/local/go/src/runtime/asm_amd64.s:2086 (0x48aef1)
```

Or do I need to put the AppImage in a directory and point butler at that?

**Comment by bjorn** [comment]
Replying to myself:
- Yes, you need to put it in a directory and push that.
- Doing so is even nice for those downloading through the web interface, since they will download a .zip and when they extract it, the file already has executable permissions. I think extracting a file from a .zip is easier for most people than adding executable permissions, so that's a win.

**Comment by fasterthanlime** [comment]
@bjorn yup! you got it. I tried Tiled on Linux while debugging downloads etc., and it just worked out of the box :) props to @probonopd and you!

Note that single-file uploads might happen sometimes in butler, but I want to make sure the special case doesn't break anything.

**Comment by probonopd** [comment]
> download a .zip and when they extract it, the file already has executable permissions

Please don't. Imagine a 6 GB game. Part of the appeal of AppImage is that you never have to unpack it, wasting no extra time or space. The optional `appimaged` daemon sets the executable bit automatically, so could the itch app.

**Comment by bjorn** [comment]
> Please don't. Imagine a 6 GB game. Part of the appeal of AppImage is that you never have to unpack it, wasting no extra time or space. The optional appimaged daemon sets the executable bit automatically, so could the itch app.

You only get a `zip` file if you use the web interface to download software, not if you use the itch app. But, you're right that this is a reason @fasterthanlime should aim to support single-file uploads for butler, and of course marking the AppImage as executable when necessary.

I also could use single-file upload support for uploading the Windows MSI installers and OS X disk image to itch, since it does not make sense to put those in a directory (not sure if the itch app would even handle that correctly?).

Btw @fasterthanlime, yesterday evening I've set up [automated `butler` pushing](https://github.com/bjorn/tiled/blob/8eec7e5210ae40869ac62d8600b421a85b678d1f/dist/linux/push-to-itch.sh) from Travis CI to itch for Tiled Linux snapshot builds. It's awesome!

**Comment by fasterthanlime** [comment]
> Please don't. Imagine a 6 GB game. Part of the appeal of AppImage is that you never have to unpack it, wasting no extra time or space. The optional appimaged daemon sets the executable bit automatically, so could the itch app.

Like @bjorn said — this is for web downloads. For the app, here's what currently happens:
- The zip is downloaded to a downloads folder
- The zip is extracted into your install folder
- The zip is removed when you uninstall the game

Here's what's going to happen soon instead:
- The zip is decompressed _while it's being downloaded_, directly into your install folder (I have working code for this, just needs more QA before releasing).

And here's what's going to happen later on:
- The uncompressed files (contents of the zip) are chopped into <=4MB blocks and compressed individually. The app downloads, uncompresses, and writes to disk (all on the fly) the ones you're missing. This also allows integrity checks and repairing existing installs.

> I also could use single-file upload support for uploading the Windows MSI installers and OS X disk image to itch, since it does not make sense to put those in a directory (not sure if the itch app would even handle that correctly?).

That would be a good use-case too! The itch app does handle nested installers (ie. installers in a .zip file) well, though!

For windows in particular, I recommend shipping a portable build — imho it's even more important than an MSI. While the app does handle MSI files, they sometimes require administrator permission to install (Tiled's does), even when the specified install folder is readable/writeable by the non-Administrator user. And there's other headaches with it: if the software's been installed outside the itch app first, the MSI install will fail (which was the whole twitter thread once the Github download issue had been solved, if you remember!). Portable builds have neither of these issues.

> Btw @fasterthanlime, yesterday evening I've set up automated butler pushing from Travis CI to itch for Tiled Linux snapshot builds. It's awesome!

So good to hear! I've made butler particularly for use cases like yours so I'm always glad to hear it works :)

**Comment by bjorn** [comment]
> The itch app does handle nested installers (ie. installers in a .zip file) well, though!

That's nice, though doing that is annoying for those downloading through the web interface. I will look into turning the OS X `dmg` into a `zip` pushed by butler though.

> For windows in particular, I recommend shipping a portable build — imho it's even more important than an MSI.

The reason I did an installer is was to add start menu entries and set up .tmx file association, and since this year also to enable the auto-update system (WinSparkle). I will consider switching to portable builds though, and maybe find some way to add a configuration for making file associations to Tiled's preferences. I see #864 will cover the start menu shortcut. :-)

Sorry for derailing this thread further, but are there any insights in the ratio of downloads from the app and downloads from the web? Intuitively, I'd expect over 90% to be downloading from the web, at least when it comes to Tiled. I'd definitely like to rely more on the update system of the app, but I don't feel like I can really push this on Tiled users as long as the app requires registration. Does the app really need that to function, or would it be conceivable to add an "unregistered" mode, where installing and updating software is still enabled?

**Comment by fasterthanlime** [comment]
> Sorry for derailing this thread further, but are there any insights in the ratio of downloads from the app and downloads from the web? Intuitively, I'd expect over 90% to be downloading from the web, at least when it comes to Tiled. I'd definitely like to rely more on the update system of the app, but I don't feel like I can really push this on Tiled users as long as the app requires registration. Does the app really need that to function, or would it be conceivable to add an "unregistered" mode, where installing and updating software is still enabled?

The app itself could be reworked to allow an anonymous mode, but there would need to be some API support for that — @leafo could probably chime in here.

**Comment by fasterthanlime** [comment]
butler v14.8.0 is out and it supports games distributed as AppImage files.

I'm not interested in distributing the app itself as an AppImage at the moment, so, closing!
